File: to-fhir/src/main/java/io/github/dizuker/tofhir/ReferenceUtils.java
package io.github.dizuker.tofhir;

import org.apache.commons.lang3.Validate;
import org.hl7.fhir.r4.model.Identifier;
import org.hl7.fhir.r4.model.Reference;
import org.hl7.fhir.r4.model.Resource;
import org.hl7.fhir.r4.model.ResourceType;

/** Utility class for creating FHIR References to resources. */
public class ReferenceUtils {
  private ReferenceUtils() {
    // Utility class, prevent instantiation
  }

  /**
   * Creates a FHIR Reference to the given resource. The reference will be in the format
   * ResourceType/Id. The resource must have a non-blank ID set.
   *
   * @param resource the FHIR resource to create a reference to
   * @return a Reference pointing to the given resource
   */
  public static Reference createReferenceTo(Resource resource) {
    Validate.notBlank(resource.getId());
    return new Reference(
        resource.getResourceType().name() + "/" + resource.getIdElement().getIdPart());
  }

  /**
   * Creates a FHIR Reference to a resource with the given identifier. The reference will be in the
   * format ResourceType/sha256(identifier.value). The identifier must have a non-blank system and
   * value set.
   *
   * @param identifier an identifier of the FHIR resource to create a reference to
   * @param resourceType the FHIR ResourceType to include in the reference
   * @return a Reference pointing to the given resource
   */
  public static Reference createReferenceTo(Identifier identifier, ResourceType resourceType) {
    return new Reference(IdUtils.fromIdentifier(identifier, resourceType));
  }
}

File: to-fhir-cs/ToFhir/ReferenceUtils.cs
using Hl7.Fhir.Model;

namespace ToFhir;

/// <summary>Utility class for creating FHIR references to resources.</summary>
public static class ReferenceUtils
{
    /// <summary>
    /// Creates a FHIR reference to the given resource, in the form <c>ResourceType/id</c>. The
    /// resource must have a non-blank <see cref="Resource.Id"/>.
    /// </summary>
    /// <param name="resource">The FHIR resource to create a reference to.</param>
    /// <returns>A reference pointing to the given resource.</returns>
    /// <exception cref="ArgumentException">The resource's id is <c>null</c>, empty or whitespace.</exception>
    public static ResourceReference CreateReferenceTo(Resource resource)
    {
        ArgumentNullException.ThrowIfNull(resource);

        if (string.IsNullOrWhiteSpace(resource.Id))
        {
            throw new ArgumentException(
                $"Resource of type {resource.TypeName} must have a non-blank id to be referenced.",
                nameof(resource)
            );
        }

        return new ResourceReference($"{resource.TypeName}/{resource.Id}");
    }

    /// <summary>
    /// Creates a FHIR reference to the resource identified by the given identifier, in the form
    /// <c>ResourceType/sha256(system|value)</c>. The identifier must have a non-blank system and
    /// value.
    /// </summary>
    /// <param name="identifier">An identifier of the FHIR resource to create a reference to.</param>
    /// <param name="resourceType">The resource type to include in the reference.</param>
    /// <returns>A reference pointing to the identified resource.</returns>
    /// <exception cref="ArgumentException">
    /// The identifier's system or value is <c>null</c>, empty or whitespace.
    /// </exception>
    public static ResourceReference CreateReferenceTo(
        Identifier identifier,
        ResourceType resourceType
    ) => new(IdUtils.FromIdentifier(identifier, resourceType));
}
